daml/Marketplace/Distribution/Auction/Service.daml

```diff
@@ -0,0 +1,201 @@
+          bidCids : [ContractId Bid.T]
+        do
+          auction@Auction{..} <- fetch auctionCid
+          assertMsg "Auction already successfully processed" (status `elem` [Auction.Open, Auction.NoValidBids])
+          bids <- forA bidCids fetch
+
+          -- Split bids on those above / below the floor price -> Sort valid bids descendingly -> Execute Auction -> Split auction result on allocated / unallocated (ie, zero allocation)
+          let
+            (validBids, invalidBids) = splitList (\bid -> bid.details.price < floorPrice) bids
+            sortedBids = sortOn (\bid -> Down bid.details) validBids
+            ((remaining, finalPrice), allocations) = mapAccumL dutchAuction (asset.quantity, 0.0) sortedBids
+            (allocatedBids, unallocatedBids) = splitList (\allocation -> allocation.quantity == 0.0) allocations
+
+          -- Return assets of all invalid and unallocated bids. Also update their bid status accordingly
+          forA_ invalidBids $ withdrawAllocation *> updateBidStatus Bid.Invalid
+          forA_ ((.bid) `map` unallocatedBids) $ withdrawAllocation *> updateBidStatus Bid.NoAllocation
+
+          -- Generate Settlement instructions for allocated bids
+          let settleAllocations = (\allocation -> Auction.SettleAllocation with price = finalPrice; issuer = customer; issuerReceivableAccount = receivableAccount; ..) <$> allocatedBids
+          (_, siCids) <- foldlA generateSettlementInstruction (Some depositCid, []) settleAllocations
+
+          -- Update Auction contract with the outcome of the auction processing
+          let newStatus = if
+                | remaining == asset.quantity -> Auction.NoValidBids
+                | remaining > 0.0             -> Auction.ParticallyAllocated with ..
+                | otherwise                   -> Auction.FullyAllocated with ..
+          auctionCid <- archive auctionCid *> create auction with status = newStatus
+
+          pure (auctionCid, siCids)
+
+template Offer
+  with
+    operator : Party
+    provider : Party
+    customer : Party
+  where
+    signatory operator, provider
+
+    controller customer can
+      Accept : ContractId Service
+        do
+          create Service with ..
+
+      Decline : ()
+        do
+          return ()
+
+template Request
+  with
+    operator : Party
+    customer : Party
+    provider : Party
+  where
+    signatory operator, customer
+
+    controller provider can
+      Approve : ContractId Service
+        do
+          create Service with ..
+
+      Reject : ()
+        do
+          return ()
+
+template CreateAuctionRequest
+  with
+    operator : Party
+    provider : Party
+    customer : Party
+    leadBidder : Party
+    asset : Asset
+    assetDepositCid : ContractId AssetDeposit
+    quotedAssetId : Id
+    allocationAccount : Account
+    receivableAccount : Account
+    floorPrice : Decimal
+  where
+    signatory operator, provider, customer
+
+template LeadBidder
```

Comment:
I'd call this something like `PublishedBid`. The issuer might decide to publicize any of the bids, doesn't have to be the lead necessarily. I think then you can also get rid of the `leadBidder` party on the auction contract.

Reply:
Done